datafusion/expr/src/type_coercion.rs

```diff
@@ -15,247 +15,49 @@
-                | UInt8
-                | UInt16
-                | UInt32
-                | UInt64
-                | Float32
-        ),
-        Float64 => matches!(
-            type_from,
-            Null | Int8
-                | Int16
-                | Int32
-                | Int64
-                | UInt8
-                | UInt16
-                | UInt32
-                | UInt64
-                | Float32
-                | Float64
-                | Decimal128(_, _)
-        ),
-        Timestamp(TimeUnit::Nanosecond, None) => {
-            matches!(type_from, Null | Timestamp(_, None))
-        }
-        Utf8 | LargeUtf8 => true,
-        Null => can_cast_types(type_from, type_into),
-        _ => false,
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use arrow::datatypes::DataType;
-
-    #[test]
-    fn test_maybe_data_types() {
-        // this vec contains: arg1, arg2, expected result
-        let cases = vec![
-            // 2 entries, same values
-            (
-                vec![DataType::UInt8, DataType::UInt16],
-                vec![DataType::UInt8, DataType::UInt16],
-                Some(vec![DataType::UInt8, DataType::UInt16]),
-            ),
-            // 2 entries, can coerse values
-            (
-                vec![DataType::UInt16, DataType::UInt16],
-                vec![DataType::UInt8, DataType::UInt16],
-                Some(vec![DataType::UInt16, DataType::UInt16]),
-            ),
-            // 0 entries, all good
-            (vec![], vec![], Some(vec![])),
-            // 2 entries, can't coerce
-            (
-                vec![DataType::Boolean, DataType::UInt16],
-                vec![DataType::UInt8, DataType::UInt16],
-                None,
-            ),
-            // u32 -> u16 is possible
-            (
-                vec![DataType::Boolean, DataType::UInt32],
-                vec![DataType::Boolean, DataType::UInt16],
-                Some(vec![DataType::Boolean, DataType::UInt32]),
-            ),
-        ];
-
-        for case in cases {
-            assert_eq!(maybe_data_types(&case.0, &case.1), case.2)
-        }
-    }
-
-    #[test]
-    fn test_get_valid_types_one_of() -> Result<()> {
-        let signature =
-            TypeSignature::OneOf(vec![TypeSignature::Any(1), TypeSignature::Any(2)]);
-
-        let invalid_types = get_valid_types(
-            &signature,
-            &[DataType::Int32, DataType::Int32, DataType::Int32],
-        )?;
-        assert_eq!(invalid_types.len(), 0);
-
-        let args = vec![DataType::Int32, DataType::Int32];
-        let valid_types = get_valid_types(&signature, &args)?;
-        assert_eq!(valid_types.len(), 1);
-        assert_eq!(valid_types[0], args);
-
-        let args = vec![DataType::Int32];
-        let valid_types = get_valid_types(&signature, &args)?;
-        assert_eq!(valid_types.len(), 1);
-        assert_eq!(valid_types[0], args);
-
-        Ok(())
-    }
-}
+pub mod aggregates;
```

Comment:
Now the code for various coercion lives in a module that helps identify what it is for